timemachine/cpp/src/smoothcore_nonbonded_interaction_group.cu

```diff
@@ -0,0 +1,391 @@
+        nblist_.resize_device(NC + NR, stream);
+
+        // Force a NBlist rebuild
+        gpuErrchk(cudaMemsetAsync(d_rebuild_nblist_, 1, 1 * sizeof(*d_rebuild_nblist_), stream));
+
+        // Offset into the ends of the arrays that now contain the row and column indices for the nblist
+        nblist_.set_idxs_device(NC, NR, d_row_atom_idxs_ + NR, d_col_atom_idxs_ + NC, stream);
+    }
+
+    // Update the row and column counts
+    this->NR_ = NR;
+    this->NC_ = NC;
+    // Reset the steps so that we do a new sort
+    this->steps_since_last_sort_ = 0;
+}
+
+template <typename RealType>
+void SmoothcoreNonbondedInteractionGroup<RealType>::du_dp_fixed_to_float(
+    const int N, const int P, const unsigned long long *du_dp, double *du_dp_float) {
+
+    for (int i = 0; i < N; i++) {
+        const int idx = i * PARAMS_PER_ATOM;
+        const int idx_charge = idx + PARAM_OFFSET_CHARGE;
+        const int idx_sig = idx + PARAM_OFFSET_SIG;
+        const int idx_eps = idx + PARAM_OFFSET_EPS;
+        const int idx_w = idx + PARAM_OFFSET_W;
+
+        du_dp_float[idx_charge] = FIXED_TO_FLOAT_DU_DP<double, FIXED_EXPONENT_DU_DCHARGE>(du_dp[idx_charge]);
+        du_dp_float[idx_sig] = FIXED_TO_FLOAT_DU_DP<double, FIXED_EXPONENT_DU_DSIG>(du_dp[idx_sig]);
+        du_dp_float[idx_eps] = FIXED_TO_FLOAT_DU_DP<double, FIXED_EXPONENT_DU_DEPS>(du_dp[idx_eps]);
+        du_dp_float[idx_w] = FIXED_TO_FLOAT_DU_DP<double, FIXED_EXPONENT_DU_DW>(du_dp[idx_w]);
+    }
+}
+
+template <typename RealType>
+void SmoothcoreNonbondedInteractionGroup<RealType>::validate_idxs(
+    const int N, const std::vector<int> &row_atom_idxs, const std::vector<int> &col_atom_idxs, const bool allow_empty) {
+
+    if (!allow_empty) {
+        if (row_atom_idxs.size() == 0) {
+            throw std::runtime_error("row_atom_idxs must be nonempty");
+        }
+        if (col_atom_idxs.size() == 0) {
+            throw std::runtime_error("col_atom_idxs must be nonempty");
+        }
+        if (row_atom_idxs.size() == static_cast<long unsigned int>(N)) {
+            throw std::runtime_error("must be less then N(" + std::to_string(N) + ") row indices");
+        }
+        if (col_atom_idxs.size() == static_cast<long unsigned int>(N)) {
+            throw std::runtime_error("must be less then N(" + std::to_string(N) + ") col indices");
+        }
+    }
+    verify_atom_idxs(N, row_atom_idxs, allow_empty);
+    verify_atom_idxs(N, col_atom_idxs, allow_empty);
+
+    // row and col idxs must be disjoint
+    std::set<int> unique_row_idxs(row_atom_idxs.begin(), row_atom_idxs.end());
+    for (int col_atom_idx : col_atom_idxs) {
+        if (unique_row_idxs.find(col_atom_idx) != unique_row_idxs.end()) {
+            throw std::runtime_error("row and col indices must be disjoint");
+        }
+    }
+}
+
+template class SmoothcoreNonbondedInteractionGroup<double>;
+template class SmoothcoreNonbondedInteractionGroup<float>;
+
+} // namespace timemachine
```

Comment:
Request: briefly describe what changed here relative to `nonbonded_interaction_group.cu`? Ideally, could create one commit that copies from `nonbonded_interaction_group.cu` and a second commit with the changes.

Reply:
Best way I can think of is to do a manual diff